src/celeritas/field/DormandPrinceMultiStepperShared.cuda.hh

```diff
@@ -0,0 +1,296 @@
+                                       int mask,
+                                       OdeState* ks,
+                                       OdeState* along_state,
+                                       FieldStepperResult* result) const;
+
+    CELER_FUNCTION void run_aside(real_type step,
+                                  OdeState const& beg_state,
+                                  int id,
+                                  int index,
+                                  int mask,
+                                  OdeState* ks,
+                                  OdeState* along_state,
+                                  FieldStepperResult* result) const;
+
+  private:
+    // Functor to calculate the force applied to a particle
+    EquationT calc_rhs_;
+};
+
+//---------------------------------------------------------------------------//
+// MACROS
+//---------------------------------------------------------------------------//
+#define UPDATE_STATE(index, state, coefficient, k)        \
+    state.pos[index - 1] = coefficient * k.pos[index - 1] \
+                           + state.pos[index - 1];        \
+    state.mom[index - 1] = coefficient * k.mom[index - 1] \
+                           + state.mom[index - 1];
+
+#define DISPATCH_VECT_MULT(mask) \
+    __syncwarp(mask);            \
+    __syncwarp(mask);
+
+//---------------------------------------------------------------------------//
+// DEDUCTION GUIDES
+//---------------------------------------------------------------------------//
+template<class EquationT>
+CELER_FUNCTION DormandPrinceMultiStepperShared(EquationT&&)
+    ->DormandPrinceMultiStepperShared<EquationT>;
+
+//---------------------------------------------------------------------------//
+// INLINE FUNCTIONS
+//---------------------------------------------------------------------------//
+#if !CELER_USE_DEVICE
+
+// TODO: implement this for CPU
+
+#endif  // !CELER_USE_DEVICE
+#ifdef __CUDA_ARCH__
+
+//---------------------------------------------------------------------------//
+// GPU FUNCTIONS
+//---------------------------------------------------------------------------//
+template<class E>
+inline CELER_FUNCTION auto
+DormandPrinceMultiStepperShared<E>::operator()(real_type step,
+                                               OdeState const& beg_state,
+                                               int number_threads) const
+    -> result_type
+{
+    int num_states = (blockDim.x * gridDim.x) / number_threads;
+    int id = (threadIdx.x + blockIdx.x * blockDim.x) / number_threads;
+    int index = (threadIdx.x + blockIdx.x * blockDim.x) % number_threads;
+
+    extern __shared__ void* shared_memory[];
+    OdeState* shared_ks = (OdeState*)shared_memory;
+    OdeState* shared_along_state
+        = reinterpret_cast<OdeState*>(&shared_ks[7 * num_states]);
+    FieldStepperResult* shared_result = reinterpret_cast<FieldStepperResult*>(
+        &shared_along_state[num_states]);
```

Comment:
You might want to add assertion to make sure that shared memory is getting indexed correctly.

Reply:
Also, are you sure that shared_result always has the correct memory alignment?